Post: "Back to Basics: commands, events, and messages" (#316)
mauroservienti/milestone.github.io: _posts/back-to-basics-messages.md

---
layout: post
title: "Back to Basics: commands, events, and messages"
author: Mauro Servienti
synopsis: "When you're involved with something for a very long time, it's easy to fall into the trap of taking many concepts for granted. Let's go back to the basics and build a common foundation. Today's topics are commands, events, and messages."
header_image: /img/posts/back-to-basics-messages/header.jpg
tags:
- messaging
- architecture
- back-to-basics
---

I often find myself talking about distributed systems and message-based architecture. Ultimately, it's been my focus for the last 15 years. I realized, though, that I often take basic concepts for granted. For example, I assume there's clarity about the differences between commands, events, and messages.

In the long term, a lack of clarity can lead to misunderstandings and suboptimal implementations that might be hard to fix.

> People used to dirt their hands with message-based architecture, and distributed systems often refer to commands and events. But what are they, and why are they so important?

Reviewer: ```suggestion
People used to dirty their hands with message-based architecture, and distributed systems often refer to commands and events. But what are they, and why are they so important?
```

When it comes to commands and events, it's all about semantics and logical concepts.

## There are only messages

Let's start with the first essential concept. Everything on the wire is a message. From the queueing system perspective, they are all messages. There is no distinction between commands and events. And that's true regardless of the infrastructure we're using. For example, Amazon Simple Notification Service (SNS) is a broadcasting service dealing with topics and subscriptions. Still, what travels on the wire are messages.

> Messages are some form of representation of a higher-level concept. They can be text messages, for example, a json or xml string, or binary messages, like a byte array. There might be restrictions, but generally, the underlying infrastructure cares very little about message format.

Reviewer: ```suggestion
Messages are some form of representation of a higher-level concept. They can be text messages, for example, or a JSON or XML string, or binary messages, like a byte array. There might be restrictions, but generally, the underlying infrastructure cares very little about message format.
```

## With messages come message headers

> More than messages alone is required. Like when we send a message in the real world, we need some surrounding metadata. For example, a text message from a mobile phone requires the recipient's phone number. A letter requires the recipient's postal address. And in both cases, the recipient receives the sender's address. The mobile phone sends it automatically, and we write our address on the envelope.

Reviewer: ```suggestion
More than messages alone is required. When we send a message in the real world, we need some surrounding metadata. For example, a text message from a mobile phone requires the recipient's phone number. A letter requires the recipient's postal address. And in both cases, the recipient receives the sender's address. The mobile phone sends it automatically, and we write our address on the envelope.
```

If different publishers could publish the same event, subscribers must inspect the incoming message to determine whether they are interested. It's a waste.

Similarly, if different recipients could handle the same command, senders would be required to have logic to determine where to send messages. It needs to be evident who the receiver is.

For those reasons, it's commonly accepted that recipients own commands and publishers own events. Recipients own commands because they own the business logic to process them. Publishers own events because they own the business logic to determine when and why to publish them.

### Why is ownership significant?

We can use ownership to determine who drives changes in the system. For example, a recipient changes a command because the processing logic changes. The changes will percolate down to senders. The opposite applies to events, publishers, and subscribers. For more information on distributed system changes and evolution, look at the [distributed systems evolution article series](https://milestone.topics.it/series/distributed-systems-evolution.html).

## Events cross service boundaries. Commands don't.

> Events are, by definition, immutable. They represent something that happened in the past. And the past cannot be changed. Events are good candidates for cross-boundaries communication.

Reviewer: ```suggestion
Events are, by definition, immutable. They represent something that happened in the past. And the past cannot be changed. Events are good candidates for cross-boundary communication.
```

> > For an introduction to services, boundaries, and components, please, refer to "[Back to Basics: service boundaries, autonomous components, and coupling](TODO-add-link)."

Author: ```suggestion
> For an introduction to services, boundaries, and components, please, refer to "[Back to Basics: service boundaries, autonomous components, and coupling](https://milestone.topics.it/2023/05/17/back-to-basics-boundaries.html)."
```

> When crossing a boundary, we're leaving a well-known territory to explore somewhere under someone else's control, a different service. We want to ensure they trust we won't move around the cheese. The second important aspect is that we have no control over someone else lawn. A service cannot tell another service, "Do this for me," it can only broadcast to those interested that something happened.

Reviewer: ```suggestion
When crossing a boundary, we're leaving a well-known territory to explore somewhere under someone else's control, a different service. We want to ensure they trust we won't move around the cheese. The second important aspect is that we have no control over someone else's lawn. A service cannot tell another service, "Do this for me," it can only broadcast to those interested that something happened.
```
